### ISSUE TITLE
Stack overflow when running `nargo execute`

### ISSUE THREAD

**Issue by huangliang-hl** [noir-lang/noir#2355]
### Aim

To successfully execute a larger neural network in the Noir programming environment.

### Expected Behavior

The neural network should compile and execute successfully, even with increased neuron counts, without encountering any errors during the 'nargo execute' command.

### Bug

I'm trying to create a larger neural network. Initially, I successfully compiled a smaller network. The model and input parameters were stored in a array, and everything worked fine.

However, when I attempted to increase the neuron counts, I encountered a error. The specific error message during the 'nargo execute' command is as follows:
```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
zsh: abort (core dumped)  /code/noir/result/bin/nargo execute
```
The code is as follows:
```
use dep::std;
global scaling_factor = 16;  
global in_channels_0 = 1;  
global out_channels_0 = 30;  
global filter_size_0 = 3;  
global image_size_0 = 28;  
global padding_0 = 1;  
global conv_stride_0 = 3;  
global image_shape_0 = [28, 28, 1];  
global filter_shape_0 = [3, 3, 1];  
global padded_image_len_0 = 900;    // (image_size+2*pading)**2*in_channels
global padded_size_0 = 30;       // image_size + 2 * padding
global convolved_image_len_0 = 3000;    // ((image_size + 2 * padding - filter_size)/conv_stride + 1)**2*out_channels
global neurons_per_layer_0 = [9, 30];  
global num_0 = 309;  
global fnn_out_len_0 = 30;  
global pooling_size_0 = 2;  
global pooling_image_len_0 = 750;      // ((convolved_size - pooling_size)/pooling_stride + 1)**2*out_channels
global pooling_stride_0 = 2;  
global convolved_size_0 = 10;    // (image_size + 2 * padding - filter_size)/conv_stride + 1
global pooled_size_0 = 5;      // (convolved_size-pooling_size)/pooling_stride+1
global pooling_kernel_len_0 = 4;    // pooling_size**2

global in_channels_1 = 30;  
global out_channels_1 = 2;  
global filter_size_1 = 3;  
global image_size_1 = 5;  
global padding_1 = 1;  
global conv_stride_1 = 1;  
global image_shape_1 = [5, 5, 30];  
global filter_shape_1 = [3, 3, 30];  
global padded_image_len_1 = 1470;    // (image_size+2*pading)**2*in_channels
global padded_size_1 = 7;       // image_size + 2 * padding
global convolved_image_len_1 = 50;    // ((image_size + 2 * padding - filter_size)/conv_stride + 1)**2*out_channels
global neurons_per_layer_1 = [270, 2];  
global num_1 = 812;  
global fnn_out_len_1 = 2;  
global pooling_size_1 = 2;  
global pooling_image_len_1 = 32;      // ((convolved_size - pooling_size)/pooling_stride + 1)**2*out_channels
global pooling_stride_1 = 1;  
global convolved_size_1 = 5;    // (image_size + 2 * padding - filter_size)/conv_stride + 1
global pooled_size_1 = 4;      // (convolved_size-pooling_size)/pooling_stride+1
global pooling_kernel_len_1 = 4;    // pooling_size**2

global neurons_per_layer_2 = [32, 10];  
global num_2 = 362;  
global fnn_out_len_2 = 10;  

global neurons_per_layer_3 = [10, 10];  
global num_3 = 120;  
global fnn_out_len_3 = 10;  

fn main(image: [u16; image_size_0*image_size_0*in_channels_0], mut inputs_0 : pub [u16; num_0], mut inputs_1 : pub [u16; num_1], mut inputs_2 : pub [u16; num_2], mut inputs_3 : pub [u16; num_3]) {

    let mut padded_image_0 : [u16; padded_image_len_0] = [0; padded_image_len_0];
    let mut convolved_image_0 : [u16; convolved_image_len_0] = [0; convolved_image_len_0];
    let mut fnn_out_0 : [u16; fnn_out_len_0] = [0; fnn_out_len_0];
    let mut pooled_image_0:[u16; pooling_image_len_0] = [0; pooling_image_len_0];
    pooled_image_0 = conv(image, inputs_0, padding_0, in_channels_0, image_size_0, padded_size_0, padded_image_0, convolved_image_0, filter_size_0, conv_stride_0, neurons_per_layer_0, fnn_out_0, fnn_out_len_0, out_channels_0, pooled_size_0, pooling_stride_0, pooling_size_0, convolved_size_0, pooled_image_0);
    std::println(pooled_image_0);

    let mut padded_image_1 : [u16; padded_image_len_1] = [0; padded_image_len_1];
    let mut convolved_image_1 : [u16; convolved_image_len_1] = [0; convolved_image_len_1];
    let mut fnn_out_1 : [u16; fnn_out_len_1] = [0; fnn_out_len_1];
    let mut pooled_image_1:[u16; pooling_image_len_1] = [0; pooling_image_len_1];
    pooled_image_1 = conv(pooled_image_0, inputs_1, padding_1, in_channels_1, image_size_1, padded_size_1, padded_image_1, convolved_image_1, filter_size_1, conv_stride_1, neurons_per_layer_1, fnn_out_1, fnn_out_len_1, out_channels_1, pooled_size_1, pooling_stride_1, pooling_size_1, convolved_size_1, pooled_image_1);
    std::println(pooled_image_1);

    inputs_2 = copy_array(pooled_image_1, inputs_2);
    let mut fnn_out_2 : [u16; fnn_out_len_2] = [0; fnn_out_len_2];
    fnn_out_2 = forward(inputs_2, neurons_per_layer_2, fnn_out_len_2, fnn_out_2);
    std::println(fnn_out_2);


    inputs_3 = copy_array(fnn_out_2, inputs_3);
    let mut fnn_out_3 : [u16; fnn_out_len_3] = [0; fnn_out_len_3];
    fnn_out_3 = forward(inputs_3, neurons_per_layer_3, fnn_out_len_3, fnn_out_3);
    std::println(fnn_out_3);


}

fn conv<M, N, P, Q, R, S, T>(image: [u16; M], mut inputs: [u16; N], padding: comptime u16, in_channels: comptime u16, image_size: comptime u16, padded_size: u16, mut padded_image: [u16; Q], mut convolved_image : [u16; R], filter_size: comptime u16, conv_stride: comptime u16, neurons_per_layer : [comptime u16; T], mut fnn_out: [u16; S], fnn_out_len: comptime u16, out_channels: comptime u16, pooled_size: comptime u16, pooling_stride: u16, pooling_size: comptime u16, convolved_size: u16, mut pooled_image: [u16; P]) -> [u16; P] {
    padded_image = pad_image(image, padding, in_channels, image_size, padded_size, padded_image);
    let step: comptime u16 = (image_size + 2 * padding - filter_size)/conv_stride + 1;
    for h_step in 0..step {
        for w_step in 0..step {
            let mut w_step_index = w_step*conv_stride;
            let mut h_step_index = h_step*conv_stride;
            for h in 0..filter_size {
                for w in 0..filter_size{
                    for c in 0..in_channels{
                        inputs[h*filter_size*in_channels+w*in_channels+c]=padded_image[(h_step_index+h)*padded_size*in_channels+(w_step_index+w)*in_channels+c];
                    }
                }
            }
            fnn_out = forward(inputs, neurons_per_layer, fnn_out_len, fnn_out);
            for n in 0..fnn_out_len {
                convolved_image[h_step*step*fnn_out_len+w_step*fnn_out_len+n] = fnn_out[n];
            }
        }
    }
    pooled_image = max_pooling(convolved_image, out_channels, pooled_size, pooling_stride, pooling_size, convolved_size, pooled_image);
    pooled_image
}

fn pad_image<M, N>(image:[u16; M], padding: comptime u16, in_channels: comptime u16, image_size: comptime u16, padded_size: u16, mut padded_image : [u16; N]) -> [u16; N] {
    // let padded_size = image_size + 2 * padding;
    for c in 0..in_channels {
        for h in 0..image_size{
            for w in 0..image_size{
                let mut padded_index = (h + padding) * padded_size * in_channels + (w + padding) * in_channels + c;
                padded_image[padded_index] = image[h * image_size * in_channels + w * in_channels + c];
            }
        }
    }
    padded_image
}

fn forward<N, M, D> (mut inputs : [u16; N], neurons_per_layer : [comptime u16; M], fnn_out_len: comptime u16, mut fnn_out : [u16; D]) -> [u16; D] {
    // inputs: [neuron0-0, neuron0-1, ..., w1-0-0, w1-0-1, ..., w1-n-0, w1-n-1, ..., b1-0, b1-1, ...,neuron1-0, neuron1-1, ..., w2-0-0, w2-0-1, ..., w2-n-0, w2-n-1, ..., b2-0, b2-1, ...]
    let mut index_layer = 0;
    for layer in 1 .. neurons_per_layer.len() {
    	index_layer += neurons_per_layer[layer - 1] * (neurons_per_layer[layer]+1);
        for i in 0 .. neurons_per_layer[layer] {
            let mut val = 0; 
            let mut index_b = index_layer + i;
            let mut index_j = index_layer - neurons_per_layer[layer - 1] * (neurons_per_layer[layer]+1);
            for j in 0 .. neurons_per_layer[layer-1] {
                let mut index_neuron = index_j+j;
                let mut index_w = index_neuron+neurons_per_layer[layer-1]*(i+1);
                val += (inputs[index_neuron]*inputs[index_w]/scaling_factor);
            }
            val += inputs[index_b];
            inputs[index_b] = activation(val);
        }
    }
    for i in 0 .. fnn_out_len {
        fnn_out[i] = inputs[inputs.len() as u16 - fnn_out_len + i];
    }
    fnn_out
}
 
fn activation(x: u16) -> u16 {
    let mut result = 0;
    if x>result {
        result = x;
    }
    result
}

fn max_pooling<M, N>(convolved_image:[u16; M], out_channels: comptime u16, pooled_size: comptime u16, pooling_stride: u16, pooling_size: comptime u16, convolved_size: u16, mut pooled_image : [u16; N]) -> [u16; N] {
    for c in 0..out_channels {
        for h_step in 0..pooled_size{
            for w_step in 0..pooled_size{
                let mut w_step_index = w_step*pooling_stride;
                let mut h_step_index = h_step*pooling_stride;
                let mut max_value: u16 = 0;
                for h in 0..pooling_size {
                    for w in 0..pooling_size{
                        let mut v=convolved_image[(h_step_index+h)*convolved_size*out_channels+(w_step_index+w)*out_channels+c];
                        if ((h==0) & (w==0)) | (v > max_value){
                            max_value = v;
                        }
                    }
                }
                let mut pooled_index = h_step * pooled_size * out_channels + w_step * out_channels + c;
                pooled_image[pooled_index] = max_value;
            }
        }
    }
    pooled_image
}

fn max<T, N>(array: [T; N]) -> T {
    let mut max_value = array[0];
    for i in 1..array.len() {
        if array[i] > max_value {
            max_value = array[i];
        }
    }
    max_value
}

fn copy_array<T, M, N>(array1: [T; M], mut array2: [T; N]) -> [T; N] {
    for i in 0..array1.len() {
        array2[i] = array1[i];
    }
    array2
}

#[test]
fn test_main() {
    let image:[u16; image_size_0*image_size_0*in_channels_0]=[0; image_size_0*image_size_0*in_channels_0];
    let inputs_0:[u16; num_0]=[0; num_0];
    let inputs_1:[u16; num_1]=[0; num_1];
    let inputs_2:[u16; num_2]=[0; num_2];
    let inputs_3:[u16; num_3]=[0; num_3];
    main(image, inputs_0, inputs_1, inputs_2, inputs_3);
}
```

### To Reproduce

1.copy the code
2.nargo test

### Installation Method

Compiled from source

### Nargo Version

nargo 0.9.0 (git version hash: 2232a387177f827d14f7c0b7ac1f3e5bb6d957d9, is dirty: false)

### Would you like to submit a PR for this Issue?

No

**Comment by kevaundray** [comment]
Hi, 

Thank you for the bug report.

When you run `nargo compile` how many opcodes are produced?

**Comment by huangliang-hl** [comment]
> Hi,
> 
> Thank you for the bug report.
> 
> When you run `nargo compile` how many opcodes are produced?

The same error is reported when running the 'nargo compile' command：
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
zsh: abort (core dumped)  /code/noir/result/bin/nargo compile

**Comment by huangliang-hl** [comment]
@kevaundray 
Hello,

Regarding this issue, are there any updates? It's a blocking issue for our project.

**Comment by jfecher** [comment]
Looks like there's a new error when running this on the most recent commit in master:
```
error: Could not determine loop bound at compile-time
   ┌─ /Users/jakefecher/Code/Noir/noir/long/src/main.nr:83:22
   │
83 │     for h_step in 0..step {
   │                      ----
   │
   = Call stack:
     1. /Users/jakefecher/Code/Noir/noir/long/src/main.nr:83:22
```
I think this may be related to #2382 which there is a PR up that should fix it (#2348), although this PR is currently blocked waiting on another PR (#2420)
